File: Free SysLog/Windows/Replacements.vb
Imports Free_SysLog.SupportCode

Public Class Replacements
    Private boolDoneLoading As Boolean = False
    Public boolChanged As Boolean = False
    Private boolEditMode As Boolean = False
    Private draggedItem As ListViewItem
    Private m_SortingColumn As ColumnHeader

    Private Sub IgnoredListView_ItemDrag(sender As Object, e As ItemDragEventArgs) Handles ReplacementsListView.ItemDrag
        draggedItem = CType(e.Item, ListViewItem)
        DoDragDrop(e.Item, DragDropEffects.Move)
    End Sub

    Private Sub IgnoredListView_DragEnter(sender As Object, e As DragEventArgs) Handles ReplacementsListView.DragEnter
        If e.Data.GetDataPresent(GetType(ListViewItem)) Then
            e.Effect = DragDropEffects.Move
        Else
            e.Effect = DragDropEffects.None
        End If
    End Sub

    Private Sub IgnoredListView_DragOver(sender As Object, e As DragEventArgs) Handles ReplacementsListView.DragOver
        e.Effect = DragDropEffects.Move
    End Sub

    Private Sub IgnoredListView_DragDrop(sender As Object, e As DragEventArgs) Handles ReplacementsListView.DragDrop
        If draggedItem Is Nothing Then Return

        Dim cp As Point = ReplacementsListView.PointToClient(New Point(e.X, e.Y))
        Dim targetItem As ListViewItem = ReplacementsListView.GetItemAt(cp.X, cp.Y)

        If targetItem Is Nothing OrElse targetItem Is draggedItem Then Return

        Dim targetIndex As Integer = targetItem.Index
        Dim draggedIndex As Integer = draggedItem.Index

        ' Remove and re-insert the dragged item
        ReplacementsListView.Items.RemoveAt(draggedIndex)
        ReplacementsListView.Items.Insert(targetIndex, draggedItem)

        ' Re-select the moved item
        draggedItem.Selected = True
        draggedItem.Focused = True

        boolChanged = True ' Mark changes
    End Sub

    Private Function CheckForExistingItem(strReplace As String, strReplaceWith As String) As Boolean
        Return ReplacementsListView.Items.Cast(Of MyReplacementsListViewItem).Any(Function(item As MyReplacementsListViewItem)
                                                                                      Return item.SubItems(0).Text.Equals(strReplace, StringComparison.OrdinalIgnoreCase) And item.SubItems(1).Text.Equals(strReplaceWith, StringComparison.OrdinalIgnoreCase)
                                                                                  End Function)
    End Function

    Private Sub BtnAdd_Click(sender As Object, e As EventArgs) Handles BtnAdd.Click
        If Not String.IsNullOrWhiteSpace(TxtReplace.Text) Then
            If ChkRegex.Checked AndAlso Not IsRegexPatternValid(TxtReplace.Text) Then
                MsgBox("Invalid regex pattern detected.", MsgBoxStyle.Critical, Text)
                Exit Sub
            End If

            If boolEditMode Then
                Dim selectedItemObject As MyReplacementsListViewItem = DirectCast(ReplacementsListView.SelectedItems(0), MyReplacementsListViewItem)

                With selectedItemObject
                    .SubItems(0).Text = TxtReplace.Text
                    .SubItems(1).Text = TxtReplaceWith.Text
                    .SubItems(2).Text = If(ChkRegex.Checked, "Yes", "No")
                    .SubItems(3).Text = If(ChkCaseSensitive.Checked, "Yes", "No")
                    .SubItems(4).Text = If(ChkEnabled.Checked, "Yes", "No")
                    .BoolRegex = ChkRegex.Checked
                    .BoolCaseSensitive = ChkCaseSensitive.Checked
                    .BoolEnabled = ChkEnabled.Checked
                    .BackColor = If(.BoolEnabled, Color.LightGreen, Color.Pink)
                End With

                ReplacementsListView.Enabled = True
                BtnAdd.Text = "Add"
                Label3.Text = "Add Replacement"
            Else
                If CheckForExistingItem(TxtReplace.Text, TxtReplaceWith.Text) Then
                    MsgBox("A similar item has already been found in your replacements list.", MsgBoxStyle.Critical, Text)
                    Exit Sub
                End If

                Dim MyReplacementsListViewItem As New MyReplacementsListViewItem(TxtReplace.Text)

                With MyReplacementsListViewItem
                    .SubItems.Add(TxtReplaceWith.Text)
                    .SubItems.Add(If(ChkRegex.Checked, "Yes", "No"))
                    .SubItems.Add(If(ChkCaseSensitive.Checked, "Yes", "No"))
                    .SubItems.Add(If(ChkEnabled.Checked, "Yes", "No"))
                    .BoolRegex = ChkRegex.Checked
                    .BoolCaseSensitive = ChkCaseSensitive.Checked
                    .BoolEnabled = ChkEnabled.Checked
                    If My.Settings.font IsNot Nothing Then .Font = My.Settings.font
                    .BackColor = If(.BoolEnabled, Color.LightGreen, Color.Pink)
                End With

                ReplacementsListView.Items.Add(MyReplacementsListViewItem)
                boolChanged = True
            End If

            boolEditMode = False
            boolChanged = True
            TxtReplace.Text = Nothing
            TxtReplaceWith.Text = Nothing
            ChkCaseSensitive.Checked = False
            ChkRegex.Checked = False
            ChkEnabled.Checked = True
        Else
            MsgBox("You need to fill in the appropriate information to create an replacement.", MsgBoxStyle.Critical, Text)
        End If
    End Sub

    Protected Overrides Sub WndProc(ByRef m As Message)
        Const WM_EXITSIZEMOVE As Integer = &H232

        MyBase.WndProc(m)

        If m.Msg = WM_EXITSIZEMOVE AndAlso boolDoneLoading Then My.Settings.replacementsLocation = Location
    End Sub

    Private Sub Replacements_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        BtnCancel.Visible = False
        Location = VerifyWindowLocation(My.Settings.replacementsLocation, Me)
        Dim listOfReplacementsToAdd As New List(Of MyReplacementsListViewItem)

        If My.Settings.replacements IsNot Nothing AndAlso My.Settings.replacements.Count > 0 Then
            For Each strJSONString As String In My.Settings.replacements
                listOfReplacementsToAdd.Add(Newtonsoft.Json.JsonConvert.DeserializeObject(Of ReplacementsClass)(strJSONString, JSONDecoderSettingsForSettingsFiles).ToListViewItem())
            Next
        End If

        ReplacementsListView.Items.AddRange(listOfReplacementsToAdd.ToArray())

        Replace.Width = My.Settings.colReplacementsReplace
        ReplaceWith.Width = My.Settings.colReplacementsReplaceWith
        Regex.Width = My.Settings.colReplacementsRegex
        CaseSensitive.Width = My.Settings.colReplacementsCaseSensitive
        ColEnabled.Width = My.Settings.colReplacementsEnabled

        Size = My.Settings.ConfigureReplacementsSize

        boolDoneLoading = True
    End Sub

    Private Sub Replacements_Closing(sender As Object, e As ComponentModel.CancelEventArgs) Handles Me.Closing
        If boolChanged Then
            Dim newReplacementsList As New ThreadSafetyLists.ThreadSafeReplacementsList
            Dim tempReplacements As New Specialized.StringCollection()
            Dim boolSuccess As Boolean = False

            Try
                Dim replacementsClass As ReplacementsClass

                For Each item As MyReplacementsListViewItem In ReplacementsListView.Items
                    replacementsClass = New ReplacementsClass With {
                        .BoolRegex = item.BoolRegex,
                        .StrReplace = item.SubItems(0).Text,
                        .StrReplaceWith = item.SubItems(1).Text,
                        .BoolCaseSensitive = item.BoolCaseSensitive,
                        .BoolEnabled = item.BoolEnabled
                    }

                    If replacementsClass.BoolEnabled Then newReplacementsList.Add(replacementsClass)
                    tempReplacements.Add(Newtonsoft.Json.JsonConvert.SerializeObject(replacementsClass))
                Next

                newReplacementsList.Sort(Function(x As ReplacementsClass, y As ReplacementsClass) x.BoolRegex.CompareTo(y.BoolRegex))
                boolSuccess = True
            Catch
            Finally
                If boolSuccess Then
                    replacementsList.Clear()
                    replacementsList.Merge(newReplacementsList)

                    My.Settings.replacements = tempReplacements
                    My.Settings.Save()
                End If
            End Try
        End If
    End Sub

    Private Sub ReplacementsListView_KeyUp(sender As Object, e As KeyEventArgs) Handles ReplacementsListView.KeyUp
        If ReplacementsListView.SelectedItems.Count > 0 Then
            If e.KeyCode = Keys.Delete Then
                BtnDelete.PerformClick()
            ElseIf e.KeyCode = Keys.Enter Then
                EditItem()
            End If
        End If
    End Sub

    Private Sub BtnDelete_Click(sender As Object, e As EventArgs) Handles BtnDelete.Click
        If ReplacementsListView.SelectedItems.Count > 0 Then
            If ReplacementsListView.SelectedItems.Count = 1 Then
                ReplacementsListView.Items.Remove(ReplacementsListView.SelectedItems(0))
            Else
                For i As Integer = ReplacementsListView.SelectedItems.Count - 1 To 0 Step -1
                    ReplacementsListView.SelectedItems(i).Remove()
                Next
            End If

            boolChanged = True
        End If
    End Sub

    Private Sub EditItem()
        If ReplacementsListView.SelectedItems.Count > 0 Then
            BtnCancel.Visible = True
            ReplacementsListView.Enabled = False
            boolEditMode = True
            BtnAdd.Text = "Save"
            Label3.Text = "Edit Replacement"

            Dim selectedItemObject As MyReplacementsListViewItem = DirectCast(ReplacementsListView.SelectedItems(0), MyReplacementsListViewItem)

            TxtReplace.Text = selectedItemObject.SubItems(0).Text
            TxtReplaceWith.Text = selectedItemObject.SubItems(1).Text
            ChkEnabled.Checked = selectedItemObject.BoolEnabled
            ChkCaseSensitive.Checked = selectedItemObject.BoolCaseSensitive
            ChkRegex.Checked = selectedItemObject.BoolRegex
        End If
    End Sub

    Private Sub BtnEdit_Click(sender As Object, e As EventArgs) Handles BtnEdit.Click
        EditItem()
    End Sub

    Private Sub ReplacementsListView_Click(sender As Object, e As EventArgs) Handles ReplacementsListView.Click
        If ReplacementsListView.SelectedItems.Count > 0 Then
            BtnDelete.Enabled = True
            BtnEdit.Enabled = True
            BtnEnableDisable.Enabled = True

            BtnEnableDisable.Text = If(DirectCast(ReplacementsListView.SelectedItems(0), MyReplacementsListViewItem).BoolEnabled, "Disable", "Enable")

            BtnUp.Enabled = ReplacementsListView.SelectedIndices(0) <> 0
            BtnDown.Enabled = ReplacementsListView.SelectedIndices(0) <> ReplacementsListView.Items.Count - 1
        Else
            BtnDelete.Enabled = False
            BtnEdit.Enabled = False
            BtnEnableDisable.Enabled = False
        End If
    End Sub

    Private Sub ReplacementsListView_DoubleClick(sender As Object, e As EventArgs) Handles ReplacementsListView.DoubleClick
        EditItem()
    End Sub

    Private Sub ListViewMenu_Opening(sender As Object, e As ComponentModel.CancelEventArgs) Handles ListViewMenu.Opening
        If ReplacementsListView.SelectedItems.Count = 0 And ReplacementsListView.SelectedItems.Count > 1 Then
            e.Cancel = True
            Exit Sub
        Else
            Dim selectedItem As MyReplacementsListViewItem = ReplacementsListView.SelectedItems(0)
            EnableDisableToolStripMenuItem.Text = If(selectedItem.BoolEnabled, "Disable", "Enable")
        End If
    End Sub

    Private Sub DisableEnableItem()
        Dim selectedItem As MyReplacementsListViewItem = ReplacementsListView.SelectedItems(0)

        If selectedItem.BoolEnabled Then
            selectedItem.BoolEnabled = False
            selectedItem.SubItems(4).Text = "No"
            BtnEnableDisable.Text = "Enable"
            selectedItem.BackColor = Color.Pink
        Else
            selectedItem.BoolEnabled = True
            selectedItem.SubItems(4).Text = "Yes"
            BtnEnableDisable.Text = "Disable"
            selectedItem.BackColor = Color.LightGreen
        End If

        boolChanged = True
    End Sub

    Private Sub BtnEnableDisable_Click(sender As Object, e As EventArgs) Handles BtnEnableDisable.Click
        DisableEnableItem()
    End Sub

    Private Sub EnableDisableToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles EnableDisableToolStripMenuItem.Click
        DisableEnableItem()
    End Sub

    Private Sub ReplacementsListView_ColumnWidthChanged(sender As Object, e As ColumnWidthChangedEventArgs) Handles ReplacementsListView.ColumnWidthChanged
        If boolDoneLoading Then
            My.Settings.colReplacementsReplace = Replace.Width
            My.Settings.colReplacementsReplaceWith = ReplaceWith.Width
            My.Settings.colReplacementsRegex = Regex.Width
            My.Settings.colReplacementsCaseSensitive = CaseSensitive.Width
            My.Settings.colReplacementsEnabled = ColEnabled.Width
        End If
    End Sub

    Private Sub Replacements_ResizeEnd(sender As Object, e As EventArgs) Handles Me.ResizeEnd
        If boolDoneLoading Then My.Settings.ConfigureReplacementsSize = Size
    End Sub

    Private Sub BtnExport_Click(sender As Object, e As EventArgs) Handles BtnExport.Click
        If ReplacementsListView.Items.Count() = 0 Then
            MsgBox("There's nothing to export.", MsgBoxStyle.Critical, Text)
            Exit Sub
        End If

        Dim saveFileDialog As New SaveFileDialog() With {.Title = "Export Replacements", .Filter = "JSON File|*.json", .OverwritePrompt = True}
        Dim listOfReplacementsClass As New List(Of ReplacementsClass)

        If saveFileDialog.ShowDialog() = DialogResult.OK Then
            For Each item As MyReplacementsListViewItem In ReplacementsListView.Items
                listOfReplacementsClass.Add(New ReplacementsClass With {.BoolRegex = item.BoolRegex, .StrReplace = item.SubItems(0).Text, .StrReplaceWith = item.SubItems(1).Text, .BoolCaseSensitive = item.BoolCaseSensitive, .BoolEnabled = item.BoolEnabled})
            Next

            IO.File.WriteAllText(saveFileDialog.FileName, Newtonsoft.Json.JsonConvert.SerializeObject(listOfReplacementsClass, Newtonsoft.Json.Formatting.Indented))

            If My.Settings.AskOpenExplorer Then
                Using OpenExplorer As New OpenExplorer()
                    OpenExplorer.StartPosition = FormStartPosition.CenterParent
                    OpenExplorer.MyParentForm = Me

                    Dim result As DialogResult = OpenExplorer.ShowDialog(Me)

                    If result = DialogResult.No Then
                        Exit Sub
                    ElseIf result = DialogResult.Yes Then
                        SelectFileInWindowsExplorer(saveFileDialog.FileName)
                    End If
                End Using
            Else
                MsgBox("Data exported successfully.", MsgBoxStyle.Information, SupportCode.ParentForm.Text)
            End If
        End If
    End Sub

    Private Sub BtnImport_Click(sender As Object, e As EventArgs) Handles BtnImport.Click
        Dim openFileDialog As New OpenFileDialog() With {.Title = "Import Alerts", .Filter = "JSON File|*.json"}
        Dim listOfReplacementsClass As New List(Of ReplacementsClass)

        If openFileDialog.ShowDialog() = DialogResult.OK Then
            Try
                listOfReplacementsClass = Newtonsoft.Json.JsonConvert.DeserializeObject(Of List(Of ReplacementsClass))(IO.File.ReadAllText(openFileDialog.FileName), JSONDecoderSettingsForLogFiles)

                ReplacementsListView.Items.Clear()
                replacementsList.Clear()

                Dim tempReplacements As New Specialized.StringCollection()

                For Each item As ReplacementsClass In listOfReplacementsClass
                    replacementsList.Add(item)
                    tempReplacements.Add(Newtonsoft.Json.JsonConvert.SerializeObject(item))
                    ReplacementsListView.Items.Add(item.ToListViewItem())
                Next

                My.Settings.replacements = tempReplacements
                My.Settings.Save()

                MsgBox("Data imported successfully.", MsgBoxStyle.Information, Text)
                boolChanged = True
            Catch ex As Newtonsoft.Json.JsonSerializationException
                MsgBox("There was an error decoding the JSON data.", MsgBoxStyle.Critical, Text)
            End Try
        End If
    End Sub

    Private Sub Replacements_KeyUp(sender As Object, e As KeyEventArgs) Handles Me.KeyUp
        If e.KeyCode = Keys.Escape Then Close()
    End Sub

    Private Sub btnDeleteAll_Click(sender As Object, e As EventArgs) Handles btnDeleteAll.Click
        If MsgBox("Are you sure you want to delete all of the replacements?", MsgBoxStyle.Question + MsgBoxStyle.YesNo, Text) = MsgBoxResult.Yes Then
            ReplacementsListView.Items.Clear()
            boolChanged = True
        End If
    End Sub

    Private Sub BtnUp_Click(sender As Object, e As EventArgs) Handles BtnUp.Click
        If ReplacementsListView.SelectedItems.Count = 0 Then Return ' No item selected
        Dim selectedIndex As Integer = ReplacementsListView.SelectedIndices(0)

        ' Ensure the item is not already at the top
        If selectedIndex > 0 Then
            Dim item As MyReplacementsListViewItem = ReplacementsListView.SelectedItems(0)
            ReplacementsListView.Items.RemoveAt(selectedIndex)
            ReplacementsListView.Items.Insert(selectedIndex - 1, item)
            ReplacementsListView.Items(selectedIndex - 1).Selected = True
            boolChanged = True
        End If

        BtnUp.Enabled = ReplacementsListView.SelectedIndices(0) <> 0
        BtnDown.Enabled = ReplacementsListView.SelectedIndices(0) <> ReplacementsListView.Items.Count - 1
    End Sub

    Private Sub BtnDown_Click(sender As Object, e As EventArgs) Handles BtnDown.Click
        If ReplacementsListView.SelectedItems.Count = 0 Then Return ' No item selected
        Dim selectedIndex As Integer = ReplacementsListView.SelectedIndices(0)

        ' Ensure the item is not already at the bottom
        If selectedIndex < ReplacementsListView.Items.Count - 1 Then
            Dim item As MyReplacementsListViewItem = ReplacementsListView.SelectedItems(0)
            ReplacementsListView.Items.RemoveAt(selectedIndex)
            ReplacementsListView.Items.Insert(selectedIndex + 1, item)
            ReplacementsListView.Items(selectedIndex + 1).Selected = True
            boolChanged = True
        End If

        BtnUp.Enabled = ReplacementsListView.SelectedIndices(0) <> 0
        BtnDown.Enabled = ReplacementsListView.SelectedIndices(0) <> ReplacementsListView.Items.Count - 1
    End Sub

    Private Sub BtnCancel_Click(sender As Object, e As EventArgs) Handles BtnCancel.Click
        ReplacementsListView.Enabled = True
        BtnAdd.Text = "Add"
        Label3.Text = "Add Replacement"
        boolEditMode = False
        boolChanged = True
        TxtReplace.Text = Nothing
        TxtReplaceWith.Text = Nothing
        ChkCaseSensitive.Checked = False
        ChkRegex.Checked = False
        ChkEnabled.Checked = True
        BtnCancel.Visible = False
    End Sub

    Private Sub IgnoredListView_ColumnClick(sender As Object, e As ColumnClickEventArgs) Handles ReplacementsListView.ColumnClick
        SortByClickedColumn(ReplacementsListView, e.Column, m_SortingColumn)
    End Sub
End Class

File: Free SysLog/Windows/Replacements.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Windows.Forms;
using Microsoft.VisualBasic;

namespace Free_SysLog
{

    public partial class Replacements
    {
        private bool boolDoneLoading = false;
        public bool boolChanged = false;
        private bool boolEditMode = false;

        public Replacements()
        {
            InitializeComponent();
        }

        private bool CheckForExistingItem(string strReplace, string strReplaceWith)
        {
            return ReplacementsListView.Items.Cast<MyReplacementsListViewItem>().Any((item) => item.SubItems[0].Text.Equals(strReplace, StringComparison.OrdinalIgnoreCase) & item.SubItems[1].Text.Equals(strReplaceWith, StringComparison.OrdinalIgnoreCase));
        }

        private void BtnAdd_Click(object sender, EventArgs e)
        {
            if (!string.IsNullOrWhiteSpace(TxtReplace.Text))
            {
                if (ChkRegex.Checked && !SupportCode.SupportCode.IsRegexPatternValid(TxtReplace.Text))
                {
                    Interaction.MsgBox("Invalid regex pattern detected.", MsgBoxStyle.Critical, Text);
                    return;
                }

                if (boolEditMode)
                {
                    MyReplacementsListViewItem selectedItemObject = (MyReplacementsListViewItem)ReplacementsListView.SelectedItems[0];

                    selectedItemObject.SubItems[0].Text = TxtReplace.Text;
                    selectedItemObject.SubItems[1].Text = TxtReplaceWith.Text;
                    selectedItemObject.SubItems[2].Text = ChkRegex.Checked ? "Yes" : "No";
                    selectedItemObject.SubItems[3].Text = ChkCaseSensitive.Checked ? "Yes" : "No";
                    selectedItemObject.SubItems[4].Text = ChkEnabled.Checked ? "Yes" : "No";
                    selectedItemObject.BoolRegex = ChkRegex.Checked;
                    selectedItemObject.BoolCaseSensitive = ChkCaseSensitive.Checked;
                    selectedItemObject.BoolEnabled = ChkEnabled.Checked;

                    ReplacementsListView.Enabled = true;
                    BtnAdd.Text = "Add";
                    Label3.Text = "Add Replacement";
                }
                else
                {
                    if (CheckForExistingItem(TxtReplace.Text, TxtReplaceWith.Text))
                    {
                        Interaction.MsgBox("A similar item has already been found in your replacements list.", MsgBoxStyle.Critical, Text);
                        return;
                    }

                    var MyReplacementsListViewItem = new MyReplacementsListViewItem(TxtReplace.Text);

                    MyReplacementsListViewItem.SubItems.Add(TxtReplaceWith.Text);
                    MyReplacementsListViewItem.SubItems.Add(ChkRegex.Checked ? "Yes" : "No");
                    MyReplacementsListViewItem.SubItems.Add(ChkCaseSensitive.Checked ? "Yes" : "No");
                    MyReplacementsListViewItem.SubItems.Add(ChkEnabled.Checked ? "Yes" : "No");
                    MyReplacementsListViewItem.BoolRegex = ChkRegex.Checked;
                    MyReplacementsListViewItem.BoolCaseSensitive = ChkCaseSensitive.Checked;
                    MyReplacementsListViewItem.BoolEnabled = ChkEnabled.Checked;
                    if (My.MySettingsProperty.Settings.font is not null)
                        MyReplacementsListViewItem.Font = My.MySettingsProperty.Settings.font;

                    ReplacementsListView.Items.Add(MyReplacementsListViewItem);
                    boolChanged = true;
                }

                boolEditMode = false;
                boolChanged = true;
                TxtReplace.Text = null;
                TxtReplaceWith.Text = null;
                ChkCaseSensitive.Checked = false;
                ChkRegex.Checked = false;
                ChkEnabled.Checked = true;
            }
            else
            {
                Interaction.MsgBox("You need to fill in the appropriate information to create an replacement.", MsgBoxStyle.Critical, Text);
            }
        }

        protected override void WndProc(ref Message m)
        {
            const int WM_EXITSIZEMOVE = 0x232;

            base.WndProc(ref m);

            if (m.Msg == WM_EXITSIZEMOVE && boolDoneLoading)
                My.MySettingsProperty.Settings.replacementsLocation = Location;
        }

        private void Replacements_Load(object sender, EventArgs e)
        {
            BtnCancel.Visible = false;
            Form argwindow = this;
            Location = SupportCode.SupportCode.VerifyWindowLocation(My.MySettingsProperty.Settings.replacementsLocation, ref argwindow);
            var listOfReplacementsToAdd = new List<MyReplacementsListViewItem>();

            if (My.MySettingsProperty.Settings.replacements is not null && My.MySettingsProperty.Settings.replacements.Count > 0)
            {
                foreach (string strJSONString in My.MySettingsProperty.Settings.replacements)
                    listOfReplacementsToAdd.Add(Newtonsoft.Json.JsonConvert.DeserializeObject<ReplacementsClass>(strJSONString, SupportCode.SupportCode.JSONDecoderSettingsForSettingsFiles).ToListViewItem());
            }

            ReplacementsListView.Items.AddRange(listOfReplacementsToAdd.ToArray());

            Replace.Width = My.MySettingsProperty.Settings.colReplacementsReplace;
            ReplaceWith.Width = My.MySettingsProperty.Settings.colReplacementsReplaceWith;
            Regex.Width = My.MySettingsProperty.Settings.colReplacementsRegex;
            CaseSensitive.Width = My.MySettingsProperty.Settings.colReplacementsCaseSensitive;
            ColEnabled.Width = My.MySettingsProperty.Settings.colReplacementsEnabled;

            Size = My.MySettingsProperty.Settings.ConfigureReplacementsSize;

            boolDoneLoading = true;
        }

        private void Replacements_Closing(object sender, System.ComponentModel.CancelEventArgs e)
        {
            if (boolChanged)
            {
                SupportCode.SupportCode.replacementsList.Clear();

                ReplacementsClass replacementsClass;
                var tempReplacements = new System.Collections.Specialized.StringCollection();

                foreach (MyReplacementsListViewItem item in ReplacementsListView.Items)
                {
                    replacementsClass = new ReplacementsClass() { BoolRegex = item.BoolRegex, StrReplace = item.SubItems[0].Text, StrReplaceWith = item.SubItems[1].Text, BoolCaseSensitive = item.BoolCaseSensitive, BoolEnabled = item.BoolEnabled };
                    if (replacementsClass.BoolEnabled)
                        SupportCode.SupportCode.replacementsList.Add(replacementsClass);
                    tempReplacements.Add(Newtonsoft.Json.JsonConvert.SerializeObject(replacementsClass));
                }

                My.MySettingsProperty.Settings.replacements = tempReplacements;
                My.MySettingsProperty.Settings.Save();
            }
        }

        private void ReplacementsListView_KeyUp(object sender, KeyEventArgs e)
        {
            if (ReplacementsListView.SelectedItems.Count > 0)
            {
                if (e.KeyCode == Keys.Delete)
                {
                    ReplacementsListView.Items.Remove(ReplacementsListView.SelectedItems[0]);
                    BtnDelete.Enabled = false;
                    BtnEdit.Enabled = false;
                    boolChanged = true;
                }
                else if (e.KeyCode == Keys.Enter)
                {
                    EditItem();
                }
            }
        }

        private void BtnDelete_Click(object sender, EventArgs e)
        {
            if (ReplacementsListView.SelectedItems.Count > 0)
            {
                if (ReplacementsListView.SelectedItems.Count == 1)
                {
                    ReplacementsListView.Items.Remove(ReplacementsListView.SelectedItems[0]);
                }
                else
                {
                    foreach (ListViewItem item in ReplacementsListView.SelectedItems)
                        item.Remove();
                }

                boolChanged = true;
            }
        }

        private void EditItem()
        {
            if (ReplacementsListView.SelectedItems.Count > 0)
            {
                BtnCancel.Visible = true;
                ReplacementsListView.Enabled = false;
                boolEditMode = true;
                BtnAdd.Text = "Save";
                Label3.Text = "Edit Replacement";

                MyReplacementsListViewItem selectedItemObject = (MyReplacementsListViewItem)ReplacementsListView.SelectedItems[0];

                TxtReplace.Text = selectedItemObject.SubItems[0].Text;
                TxtReplaceWith.Text = selectedItemObject.SubItems[1].Text;
                ChkEnabled.Checked = selectedItemObject.BoolEnabled;
                ChkCaseSensitive.Checked = selectedItemObject.BoolCaseSensitive;
                ChkRegex.Checked = selectedItemObject.BoolRegex;
            }
        }

        private void BtnEdit_Click(object sender, EventArgs e)
        {
            EditItem();
        }

        private void ReplacementsListView_Click(object sender, EventArgs e)
        {
            if (ReplacementsListView.SelectedItems.Count > 0)
            {
                BtnDelete.Enabled = true;
                BtnEdit.Enabled = true;
                BtnEnableDisable.Enabled = true;

                BtnEnableDisable.Text = ((MyReplacementsListViewItem)ReplacementsListView.SelectedItems[0]).BoolEnabled ? "Disable" : "Enable";

                BtnUp.Enabled = ReplacementsListView.SelectedIndices[0] != 0;
                BtnDown.Enabled = ReplacementsListView.SelectedIndices[0] != ReplacementsListView.Items.Count - 1;
            }
            else
            {
                BtnDelete.Enabled = false;
                BtnEdit.Enabled = false;
                BtnEnableDisable.Enabled = false;
            }
        }

        private void ReplacementsListView_DoubleClick(object sender, EventArgs e)
        {
            EditItem();
        }

        private void ListViewMenu_Opening(object sender, System.ComponentModel.CancelEventArgs e)
        {
            if (ReplacementsListView.SelectedItems.Count == 0 & ReplacementsListView.SelectedItems.Count > 1)
            {
                e.Cancel = true;
                return;
            }
            else
            {
                MyReplacementsListViewItem selectedItem = (MyReplacementsListViewItem)ReplacementsListView.SelectedItems[0];
                EnableDisableToolStripMenuItem.Text = selectedItem.BoolEnabled ? "Disable" : "Enable";
            }
        }

        private void DisableEnableItem()
        {
            MyReplacementsListViewItem selectedItem = (MyReplacementsListViewItem)ReplacementsListView.SelectedItems[0];

            if (selectedItem.BoolEnabled)
            {
                selectedItem.BoolEnabled = false;
                selectedItem.SubItems[4].Text = "No";
                BtnEnableDisable.Text = "Enable";
            }
            else
            {
                selectedItem.BoolEnabled = true;
                selectedItem.SubItems[4].Text = "Yes";
                BtnEnableDisable.Text = "Disable";
            }

            boolChanged = true;
        }

        private void BtnEnableDisable_Click(object sender, EventArgs e)
        {
            DisableEnableItem();
        }

        private void EnableDisableToolStripMenuItem_Click(object sender, EventArgs e)
        {
            DisableEnableItem();
        }

        private void ReplacementsListView_ColumnWidthChanged(object sender, ColumnWidthChangedEventArgs e)
        {
            if (boolDoneLoading)
            {
                My.MySettingsProperty.Settings.colReplacementsReplace = Replace.Width;
                My.MySettingsProperty.Settings.colReplacementsReplaceWith = ReplaceWith.Width;
                My.MySettingsProperty.Settings.colReplacementsRegex = Regex.Width;
                My.MySettingsProperty.Settings.colReplacementsCaseSensitive = CaseSensitive.Width;
                My.MySettingsProperty.Settings.colReplacementsEnabled = ColEnabled.Width;
            }
        }

        private void Replacements_ResizeEnd(object sender, EventArgs e)
        {
            if (boolDoneLoading)
                My.MySettingsProperty.Settings.ConfigureReplacementsSize = Size;
        }

        private void BtnExport_Click(object sender, EventArgs e)
        {
            if (ReplacementsListView.Items.Count == 0)
            {
                Interaction.MsgBox("There's nothing to export.", MsgBoxStyle.Critical, Text);
                return;
            }

            var saveFileDialog = new SaveFileDialog() { Title = "Export Replacements", Filter = "JSON File|*.json", OverwritePrompt = true };
            var listOfReplacementsClass = new List<ReplacementsClass>();

            if (saveFileDialog.ShowDialog() == DialogResult.OK)
            {
                foreach (MyReplacementsListViewItem item in ReplacementsListView.Items)
                    listOfReplacementsClass.Add(new ReplacementsClass() { BoolRegex = item.BoolRegex, StrReplace = item.SubItems[0].Text, StrReplaceWith = item.SubItems[1].Text, BoolCaseSensitive = item.BoolCaseSensitive, BoolEnabled = item.BoolEnabled });

                System.IO.File.WriteAllText(saveFileDialog.FileName, Newtonsoft.Json.JsonConvert.SerializeObject(listOfReplacementsClass, Newtonsoft.Json.Formatting.Indented));

                Interaction.MsgBox("Data exported successfully.", MsgBoxStyle.Information, Text);
            }
        }

        private void BtnImport_Click(object sender, EventArgs e)
        {
            var openFileDialog = new OpenFileDialog() { Title = "Import Alerts", Filter = "JSON File|*.json" };
            var listOfReplacementsClass = new List<ReplacementsClass>();

            if (openFileDialog.ShowDialog() == DialogResult.OK)
            {
                try
                {
                    listOfReplacementsClass = Newtonsoft.Json.JsonConvert.DeserializeObject<List<ReplacementsClass>>(System.IO.File.ReadAllText(openFileDialog.FileName), SupportCode.SupportCode.JSONDecoderSettingsForLogFiles);

                    ReplacementsListView.Items.Clear();
                    SupportCode.SupportCode.replacementsList.Clear();

                    var tempReplacements = new System.Collections.Specialized.StringCollection();

                    foreach (ReplacementsClass item in listOfReplacementsClass)
                    {
                        SupportCode.SupportCode.replacementsList.Add(item);
                        tempReplacements.Add(Newtonsoft.Json.JsonConvert.SerializeObject(item));
                        ReplacementsListView.Items.Add(item.ToListViewItem());
                    }

                    My.MySettingsProperty.Settings.replacements = tempReplacements;
                    My.MySettingsProperty.Settings.Save();

                    Interaction.MsgBox("Data imported successfully.", MsgBoxStyle.Information, Text);
                    boolChanged = true;
                }
                catch (Newtonsoft.Json.JsonSerializationException)
                {
                    Interaction.MsgBox("There was an error decoding the JSON data.", MsgBoxStyle.Critical, Text);
                }
            }
        }

        private void Replacements_KeyUp(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Escape)
                Close();
        }

        private void btnDeleteAll_Click(object sender, EventArgs e)
        {
            if (Interaction.MsgBox("Are you sure you want to delete all of the replacements?", (MsgBoxStyle)((int)MsgBoxStyle.Question + (int)MsgBoxStyle.YesNo), Text) == MsgBoxResult.Yes)
            {
                ReplacementsListView.Items.Clear();
                boolChanged = true;
            }
        }

        private void BtnUp_Click(object sender, EventArgs e)
        {
            if (ReplacementsListView.SelectedItems.Count == 0)
                return; // No item selected
            int selectedIndex = ReplacementsListView.SelectedIndices[0];

            // Ensure the item is not already at the top
            if (selectedIndex > 0)
            {
                MyReplacementsListViewItem item = (MyReplacementsListViewItem)ReplacementsListView.SelectedItems[0];
                ReplacementsListView.Items.RemoveAt(selectedIndex);
                ReplacementsListView.Items.Insert(selectedIndex - 1, item);
                ReplacementsListView.Items[selectedIndex - 1].Selected = true;
                boolChanged = true;
            }

            BtnUp.Enabled = ReplacementsListView.SelectedIndices[0] != 0;
            BtnDown.Enabled = ReplacementsListView.SelectedIndices[0] != ReplacementsListView.Items.Count - 1;
        }

        private void BtnDown_Click(object sender, EventArgs e)
        {
            if (ReplacementsListView.SelectedItems.Count == 0)
                return; // No item selected
            int selectedIndex = ReplacementsListView.SelectedIndices[0];

            // Ensure the item is not already at the bottom
            if (selectedIndex < ReplacementsListView.Items.Count - 1)
            {
                MyReplacementsListViewItem item = (MyReplacementsListViewItem)ReplacementsListView.SelectedItems[0];
                ReplacementsListView.Items.RemoveAt(selectedIndex);
                ReplacementsListView.Items.Insert(selectedIndex + 1, item);
                ReplacementsListView.Items[selectedIndex + 1].Selected = true;
                boolChanged = true;
            }

            BtnUp.Enabled = ReplacementsListView.SelectedIndices[0] != 0;
            BtnDown.Enabled = ReplacementsListView.SelectedIndices[0] != ReplacementsListView.Items.Count - 1;
        }

        private void BtnCancel_Click(object sender, EventArgs e)
        {
            ReplacementsListView.Enabled = true;
            BtnAdd.Text = "Add";
            Label3.Text = "Add Replacement";
            boolEditMode = false;
            boolChanged = true;
            TxtReplace.Text = null;
            TxtReplaceWith.Text = null;
            ChkCaseSensitive.Checked = false;
            ChkRegex.Checked = false;
            ChkEnabled.Checked = true;
            BtnCancel.Visible = false;
        }
    }
}